### PR TITLE
PLT-5551 Unit Tests [PLT-5867 + PLT-5868]

### PR DESCRIPTION
**Pull request by amnambiar**
https://input-output.atlassian.net/browse/PLT-5551

## Release Notes

- **New Feature**: Added a new route `CreateAuditorReport` for creating auditor reports.
- **Refactor**: Updated the `Counter` class in Python code to be thread-safe, preventing data races.
- **Test**: Enhanced testing capabilities by adding `data-testid` attributes to various React components and introducing new test cases.
- **Bug Fix**: Corrected the regular expression pattern in `parseHTMLContents` function to match closing `</h2>` tag correctly.

> 🎉 Celebrate this PR, it's no small feat,  
> New routes and tests, oh so neat!  
> Thread-safe counters, bugs retreat,  
> Code now stronger, can't be beat! 🚀

@coderabbitai: ignore

### CODE REVIEW

**Comment by amnambiar** [comment]
There are some failures even at this point, but I would recommend to review the setup of the tests, and the way they are written for the time-being. As, most of them will definitely have to be rewritten once the revamp is in place.

**Comment by RSoulatIOHK** [comment]
> Image description CodeRabbit
> 
> Commits
> Files selected for processing (1)
> Files skipped from review due to trivial changes (1)
> Review comments generated (0)
> Tips

@coderabbitai: Can you review the file that you skipped?

**Review by RSoulatIOHK — COMMENTED** [review]
@coderabbit Can you review the latest commit?

File: react-web/src/pages/certification/components/SuccessCard.test.tsx [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    };
+
+    render(
+      <SuccessCard {...mockData} />
+    );
+    expect(screen.getByText(/Task: Task 1/)).toBeInTheDocument();
+    expect(screen.queryByTestId("_certRes_DLTests")).not.toBeInTheDocument();
+    expect(screen.queryByTestId("_certRes_unitTestResults")).not.toBeInTheDocument();
+  });
+});
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply react-web/src/pages/certification/components/SuccessCard.test.tsx:218]: @coderabbitai Can you also review the file in the latest commit, SuccessCard.tsx?

**Review by bogdan-manole — APPROVED** [review]

File: react-web/src/components/Icons/Icons.test.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    );
+
+    const iconElement = container.querySelector(".MuiSvgIcon-root");
+    expect(iconElement).toBeInTheDocument();
+
+    // Simulate a click on the icon
+    fireEvent.click(iconElement!);
+    expect(onClick).toHaveBeenCalled();
+  });
+});
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply react-web/src/components/Icons/Icons.test.tsx:54]: ignoring this as it throws other type ts error , and deciding to keep it as such

File: react-web/src/compositions/Timeline/Timeline.test.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        unitTestSuccess={unitTestSuccess}
+        hasFailedTasks={hasFailedTasks}
+        buildInfo={buildInfo}
+      />
+    );
+
+    // Assert that the runTimeTaken property is set correctly for the running status
+    const runningTimelineItem = screen.getByTestId("undefined-runTimeTaken");
+    expect(runningTimelineItem.textContent).toBe("1000");
+  });
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply react-web/src/compositions/Timeline/Timeline.test.tsx:65]: Ignoring now, will be looked at while revamp

File: react-web/src/pages/certification/components/FailedStatus.test.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  });
+
+  test("toggles accordion on click", () => {
+    render(<FailedStatus {...mockData} />);
+
+    expect(screen.getByTestId("arrow")).toHaveClass("up");
+    fireEvent.click(screen.getByTestId("accordion-title"));
+    expect(screen.getByTestId("arrow")).toHaveClass("down");
+  });
+});
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply react-web/src/pages/certification/components/FailedStatus.test.tsx:44]: ignoring this for now

File: react-web/src/pages/certification/components/SuccessCard.test.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (amnambiar) [reply react-web/src/pages/certification/components/SuccessCard.test.tsx:218]: ignoring this for now

File: react-web/src/pages/certification/components/ResultContainer.test.tsx [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    const { queryByTestId, queryByText } = render(
+      <ResultContainer {...mockData} unitTestSuccess={false} />
+    );
+
+    expect(queryByTestId("failure-output")).toBeInTheDocument();
+    expect(
+      queryByText(/InstanceErr (GenericError "failed to create currency")/i)
+    ).not.toBeInTheDocument();
+  });
+});
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply react-web/src/pages/certification/components/ResultContainer.test.tsx:156]: ignoring for now due to type errors

File: react-web/src/components/Table/components/ColViz/ColViz.test.tsx [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    );
+
+    const columnOption = screen.getByTestId(`${columns[1].Header}-checkbox`);
+    fireEvent.click(columnOption);
+
+    const checkbox = screen.getByTestId(`${columns[0].Header}-checkbox`).firstChild as HTMLInputElement;
+
+    expect(checkbox).toBeDisabled();
+  });
+});
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply react-web/src/components/Table/components/ColViz/ColViz.test.tsx:200]: Will be looked at later as running short of time

File: react-web/src/compositions/Timeline/Timeline.test.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (amnambiar) [reply react-web/src/compositions/Timeline/Timeline.test.tsx:65]: Unable to do this quickly.Will be looked at later as running short of time

File: react-web/src/compositions/Timeline/Timeline.test.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      <Timeline
+        statusConfig={statusConfig}
+        unitTestSuccess={unitTestSuccess}
+        hasFailedTasks={hasFailedTasks}
+      />
+    );
+
+    const timelineItems = screen.getAllByRole("listitem");
+    expect(timelineItems).toHaveLength(statusConfig.length);
+  });
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply react-web/src/compositions/Timeline/Timeline.test.tsx:38]: Will be looked at later as running short of time

File: react-web/src/components/Table/components/ColViz/ColViz.tsx [hunk 8]
```diff
@@ -77,7 +80,7 @@ const ColViz: FC<any> = ({ columns, updateColumnOptions }) => {
               </li>
             );
           } else {
-            return <></>
+            return <span key={uniqueId("colviz")}></span>
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply react-web/src/components/Table/components/ColViz/ColViz.tsx:83]: Need to recheck the implementation.Will do that later as running short of time

File: react-web/src/components/Table/components/ColViz/ColViz.tsx [hunk 9]
```diff
@@ -56,19 +58,20 @@ const ColViz: FC<any> = ({ columns, updateColumnOptions }) => {
...
-              <li key={index}>
+              <li key={uniqueId("colviz")}>
                 <FormGroup>
                   <FormControlLabel
                     control={
                       <Checkbox
                         checked={column.columnVisible}
                         onChange={(e) => handleChange(e, column.accessor)}
                         disabled={column.checkBoxDisabled}
+                        data-testid={`${column?.Header}-checkbox`}
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply react-web/src/components/Table/components/ColViz/ColViz.tsx:74]: Need to recheck the implementation.Will do that later as running short of time

File: react-web/src/pages/certification/Certification.test.tsx [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+    });
+
+    jest.advanceTimersByTime(5000);
+
+    // screen.debug();
+
+    // When form is submitting
+    expect(screen.queryByTestId("resultContainer")).toBeInTheDocument();
+    expect(screen.queryByTestId("abort")).toBeInTheDocument();
+  });
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply react-web/src/pages/certification/Certification.test.tsx:278]: Will work later on this as running short of time

File: react-web/src/utils/utils.ts [hunk 11]
```diff
@@ -56,6 +56,11 @@ export const formatTimeToReadable = (duration: number) => {
     return timeStr
 }
 
+// unused now
+// export const convertAdaToLovelace = (fee_ada: number) => {
+//   return BigNum.from_str((fee_ada * 1000000).toString())
+// }
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply react-web/src/utils/utils.ts:62]: Keeping it as such for now. Will do this later

File: react-web/src/utils/utils.ts [hunk 12]
```diff
@@ -56,6 +56,11 @@ export const formatTimeToReadable = (duration: number) => {
...
+
 export const getObjectByPath = (object: { [x: string]: any }, path: string): any => {
   path = path.replace(/\[(\w+)\]/g, ".$1"); // convert indexes to properties
   path = path.replace(/^\./, ""); // strip a leading dot
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply react-web/src/utils/utils.ts:66]: ignoring this for now

**Comment by amnambiar** [comment]
Have couple test failures which will be taken care of during the revamp tickets. Merging this PR for now.

**Pull request merged**